### PR TITLE
support for federated clusters

### LINKED ISSUE #14535: support for federated clusters
### Motivation

Data centers may be located in different regions and provided by different cloud providers. Usually, dedicated network connections are used to interconnect the networks of these data centers, facilitating data transmission. However, transferring data between data centers incurs expensive bandwidth costs and significant transmission delays. Therefore, it is essential to minimize data transfer between different data centers as much as possible. Rather than centralizing data storage in a single data center for all regions, we prefer to report data to the nearest data center. However, this approach poses certain challenges. If there are a large number of data centers, maintaining numerous independent Druid clusters becomes necessary. As these clusters operate independently, it can result in inaccurate aggregated queries. By enabling support for federated clusters in Druid databases, substantial cost savings in terms of bandwidth can be achieved.

### Description

Here is one possible deployment architecture:

image

To enable federated queries, the "federatedClusterBrokers" property should be added to the context. Here is an example query:
```
curl 'http://localhost:8888/druid/v2/sql' \
  -H 'sec-ch-ua: "Not.A/Brand";v="8", "Chromium";v="114", "Google Chrome";v="114"' \
  -H 'Accept: application/json, text/plain, */*' \
  -H 'Content-Type: application/json' \
  -H 'Referer: http://localhost:8888/unified-console.html' \
  --data-raw '{"query":"SELECT COUNT(*),APPROX_QUANTILE_DS(added, 0.95) \"P95cdnTime\",page FROM wikipedia GROUP BY page ORDER BY 1 DESC limit 10","resultFormat":"array","header":true,"typesHeader":true,"sqlTypesHeader":true,"context":{"sqlOuterLimit":1001,"federatedClusterBrokers":"sanFrancisco:8888,hongKong:8888","sqlQueryId":"7e7c37e7-1cd4-49c2-9a20-8950350b7997"}}' 
```

The above is the current approach I am using to implement a federated cluster with Druid. Are there any better ways to support federated queries in Druid clusters?

### PR DESCRIPTION
**Pull request by 599166320**
Fixes #14535.

### Description

1. Druid provides a friendly and unified query gateway for users across multiple data centers and clusters through cluster federation queries.

2. As the number of nodes and metadata increases, a single Druid cluster can become excessively large, leading to mutual interference among tasks and deteriorating scheduling performance. Implementing federated queries can help avoid such issues. It allows for breaking down large clusters into relatively independent ones as needed, making scheduling more agile and lightweight.

<hr>

##### Key changed/added classes in this PR
 * `QueryContexts`
 * `BrokerServerView`
 * `CachingClusteredClient`
 * `TimelineServerView`

<hr>

This PR has:

- [x] been self-reviewed.
   - [x] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: processing/src/main/java/org/apache/druid/query/QueryContexts.java [hunk 1]
```diff
@@ -85,6 +85,7 @@ public class QueryContexts
   public static final String SERIALIZE_DATE_TIME_AS_LONG_INNER_KEY = "serializeDateTimeAsLongInner";
   public static final String UNCOVERED_INTERVALS_LIMIT_KEY = "uncoveredIntervalsLimit";
   public static final String MIN_TOP_N_THRESHOLD = "minTopNThreshold";
+  public static final String FEDERATED_CLUSSTER_BROKERS = "federatedClusterBrokers";
```
Comment (funguy-tech) [inline processing/src/main/java/org/apache/druid/query/QueryContexts.java:88]: Nit: Should this be `FEDERATED_CLUSTER_BROKERS` instead?

**Comment by cryptoe** [comment]
@599166320 
How big is the deployment you are working witch is causing things like 
```
 leading to mutual interference among tasks and deteriorating scheduling performance.
 ```
 I really donot understand what this means. Are you talking about ingestion tasks or query scheduling. 
 
The change in the current form looks hackish . 
I think its a better design pattern that one cluster owns one data source if you really want to break things up since then you can configure load/rules compaction etc only on one cluster for a data source. Ingestion also gets simpler.

**Comment by 599166320** [comment]
@cryptoe 
Thank you for your response. For the second point, I also hesitated whether to include it. However, I decided to include it to see what everyone thinks.

Usually, we deploy a Druid cluster in one data center, with approximately a hundred servers in each data center. The entire cluster is quite stable. However, we've noticed that the master node operates in a master-slave configuration, storing a significant amount of metadata and handling heavy scheduling tasks. In theory, there might be bottlenecks, so we wanted to bring it up for discussion.

Of course, what we are more concerned about is the cost of dedicated network traffic.

**Comment by 599166320** [comment]
@cryptoe 
Is there a better way to implement the practical application of federated queries?  @abhishekagarwal87 mentioned that it might potentially break certain protocols. In my opinion, the native query for Historical and Broker currently shares the code using QueryResource as the entry point, and the parameter structures are almost identical.

However, if strict constraints are necessary, Query.queryContext might require some improvements. What are your thoughts on this?

**Comment by cryptoe** [comment]
I think the correct way to do it would be to use something like https://github.com/lyft/presto-gateway and pass a query context to select the correct cluster you want or make some mapping to data source -> cluster on this gateway nodes.

**Comment by cryptoe** [comment]
Another thing I was thinking is how order by's would look. The broker expects things to be sorted by the grouping key and then sorts stuff on the order by key IIRC. 
In this case the cluster 2 broker will return the rows already sorted on the order by key, which will break the merging logic of the grouping keys on the broker for cluster 1.

**Comment by 599166320** [comment]
@cryptoe According to your description, are you concerned about a sorting like the one below?
```
SELECT
  COUNT(*) c,
  regionName
FROM wikipedia
GROUP BY regionName
ORDER BY regionName, c DESC
LIMIT 10

```
In fact, when a query like this is forwarded from one broker to another broker in a cluster, the `LIMIT 10` part is removed. It will be transformed into a `native query `similar to the one below:
```
{
    "queryType": "groupBy",
    "dataSource":
    {
        "type": "table",
        "name": "wikipedia"
    },
    "intervals":
    {
        "type": "intervals",
        "intervals":
        [
            "-146136543-09-08T08:23:32.096Z/146140482-04-24T15:36:27.903Z"
        ]
    },
    "granularity":
    {
        "type": "all"
    },
    "dimensions":
    [
        {
            "type": "default",
            "dimension": "regionName",
            "outputName": "d0",
            "outputType": "STRING"
        }
    ],
    "aggregations":
    [
        {
            "type": "count",
            "name": "a0"
        }
    ],
    "limitSpec":
    {
        "type": "NoopLimitSpec"
    },
    "context":
    {
        "applyLimitPushDown": false,
        "defaultTimeout": 300000,
        "federatedClusterBrokers": "",
        "finalize": false,
        "fudgeTimestamp": "-4611686018427387904",
        "groupByOutermost": false,
        "groupByStrategy": "v2",
        "maxQueuedBytes": 5242880,
        "maxScatterGatherBytes": 9223372036854775807,
        "queryFailTime": 1699369214285,
        "queryId": "605a751b-f0ee-43fe-a754-b702035622df",
        "resultAsArray": true,
        "sqlQueryId": "e8bed46a-f2de-4fc3-89c6-febfe048debc",
        "timeout": 29981
    }
}

```
So, you don't need to worry about this sorting issue.

**Pull request closed without merging**